frame/two-phase-election-provider/src/lib.rs

```diff
@@ -0,0 +1,1627 @@
+					},
+					Err(why) => log!(error, "Error in unsigned offchain worker: {:?}", why),
+				}
+			}
+		}
+
+		fn integrity_test() {
+			let max_vote: usize = <CompactOf<T> as CompactSolution>::LIMIT;
+			// based on the requirements of [`sp_npos_elections::Assignment::try_normalize`].
+
+			// 1. Maximum sum of [ChainAccuracy; 16] must fit into `UpperOf<ChainAccuracy>`..
+			let maximum_chain_accuracy: Vec<UpperOf<OnChainAccuracyOf<T>>> = (0..max_vote)
+				.map(|_| <OnChainAccuracyOf<T>>::one().deconstruct().into())
+				.collect();
+			let _: UpperOf<OnChainAccuracyOf<T>> = maximum_chain_accuracy
+				.iter()
+				.fold(Zero::zero(), |acc, x| acc.checked_add(x).unwrap());
+
+			// 2. Maximum sum of [CompactAccuracy; 16] must fit into `UpperOf<OffchainAccuracy>`.
+			let maximum_chain_accuracy: Vec<UpperOf<CompactAccuracyOf<T>>> = (0..max_vote)
+				.map(|_| <CompactAccuracyOf<T>>::one().deconstruct().into())
+				.collect();
+			let _: UpperOf<CompactAccuracyOf<T>> = maximum_chain_accuracy
+				.iter()
+				.fold(Zero::zero(), |acc, x| acc.checked_add(x).unwrap());
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		ExtendedBalance: From<InnerOf<CompactAccuracyOf<T>>>,
+		ExtendedBalance: From<InnerOf<OnChainAccuracyOf<T>>>,
+	{
+		/// Submit a solution for the signed phase.
+		///
+		/// The dispatch origin fo this call must be __signed__.
+		///
+		/// The solution is potentially queued, based on the claimed score and processed at the end
+		/// of the signed phase.
+		///
+		/// A deposit is reserved and recorded for the solution. Based on the outcome, the solution
+		/// might be rewarded, slashed, or get all or a part of the deposit back.
+		///
+		/// # <weight>
+		/// Queue size must be provided as witness data.
+		/// # </weight>
+		#[pallet::weight(T::WeightInfo::submit(*witness_data))]
+		pub fn submit(
+			origin: OriginFor<T>,
+			solution: RawSolution<CompactOf<T>>,
+			witness_data: u32,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			// ensure witness data is correct.
+			ensure!(
+				witness_data >= <SignedSubmissions<T>>::decode_len().unwrap_or_default() as u32,
+				Error::<T>::InvalidWitness,
+			);
+
+			// ensure solution is timely.
+			ensure!(
+				Self::current_phase().is_signed(),
+				Error::<T>::EarlySubmission,
+			);
+
+			// NOTE: this is the only case where having separate snapshot would have been better
+			// because could do just decode_len. But we can create abstractions to do this.
+
+			// build size. Note: this is not needed for weight calc, thus not input.
+			// defensive-only: if phase is signed, snapshot will exist.
+			let size = Self::build_solution_size().unwrap_or_default();
+
+			// ensure solution claims is better.
+			let mut signed_submissions = Self::signed_submissions();
+			let maybe_index =
+				Self::insert_submission(&who, &mut signed_submissions, solution, size);
```

Comment:
This is a storage mutating call.

insert_submission -> remove_weakest -> unreserve

Then you cannot have the ensure fail below

Reply:
Note that the call path happens only when it is followed by returning `Some`, which means that the asseriton in the next line will not fail. 

Nonetheless, bad design and shall change. You shall receive a gift from the Parity beer fund for your pedantic eye for such mistakes 💯

Reply:
With a little bit of fiddling around.. I actually like the fact that the function does the removal as well and always keep the queue size full. So will merge the lines to sth like 

```
let index = Self::insert_submission(&who, &mut signed_submissions, solution, size).ok_or(...)?;
```

Conceptually the same thing, but I think it will avoid confusion.

Reply:
but for example if the sender cannot pay for the deposit then, the weakest has been unreserved while still in the queue

Reply:
ahh I didn't see the reserve operation further down 🤦  yes that's wrong.

Reply:
addressed.